src/python/deepgnn/graph_engine/snark/dispatcher.py

```diff
@@ -151,11 +150,53 @@ def join(self):
             flag, output = self.q_out.get()
             self.node_count += output["node_count"]
             self.edge_count += output["edge_count"]
+            self.node_type_num = max(self.node_type_num, output["node_type_num"])
+            self.edge_type_num = max(self.edge_type_num, output["edge_type_num"])
+            self.node_feature_num = max(
+                self.node_feature_num, output["node_feature_num"]
+            )
+            self.edge_feature_num = max(
+                self.edge_feature_num, output["edge_feature_num"]
+            )
             self.partitions.append(output["partition"])
 
             assert flag == FLAG_WORKER_FINISHED_PROCESSING
         self.q_out.close()
 
+        for p in self.partitions:
+            real_node_type_num = len(p["node_type_count"])
+            if real_node_type_num < self.node_type_num:
+                p["node_type_count"].extend(
+                    [0] * (self.node_type_num - real_node_type_num)
+                )
+                p["node_weight"].extend(
+                    [0.0] * (self.node_type_num - real_node_type_num)
+                )
+            real_edge_type_num = len(p["edge_type_count"])
+            if real_edge_type_num < self.edge_type_num:
+                p["edge_type_count"].extend(
+                    [0] * (self.edge_type_num - real_edge_type_num)
+                )
+                p["edge_weight"].extend([0] * (self.edge_type_num - real_edge_type_num))
+
+            fs, _ = get_fs(self.folder)
+            for tp in range(real_node_type_num, self.node_type_num):
```

Comment:
1. This code seems to always create alias tables, but users can skip them. Do we have to create empty files, can we avoid it?
2. Alias tables are sharded by partition id I think, so range should start from 0 instead of `real_node_type_num`.

Reply:
1. Done.
2. They are but real_node_type_num = len(p["node_type_count"]) so it is per partition.